### PR TITLE
Fixes an issue with XSuite not being able to run some formulations of XQsuite

### PR DESCRIPTION
**Pull request by adamretter**
Closes https://github.com/eXist-db/exist/issues/4588

### CODE REVIEW

**Review by dizzzz — APPROVED** [review]

**Comment by line-o** [comment]
Can you clarify what you mean by XSuite is it XQSuite? Can those terms be used interchangeably?

**Review by reinhapa — APPROVED** [review]

File: exist-core/src/test/resources-filtered/conf.xml [hunk 1]
```diff
@@ -888,6 +888,10 @@
             raise-error-on-failed-retrieval="no">
         
         <builtin-modules>
+
+            <!-- XQSuite Test Framework for XQuery testing -->
```
Comment (line-o) [inline exist-core/src/test/resources-filtered/conf.xml:892]: Will this now effectively add XQSuite to the XQuery contexts of the next release?

File: exist-core/src/main/java/org/exist/xquery/ModuleContext.java [hunk 2]
```diff
@@ -65,14 +65,15 @@ public class ModuleContext extends XQueryContext {
 
     public ModuleContext(final XQueryContext parentContext, final String modulePrefix, final String moduleNamespace,
             final String location) {
-        super();
+        super(parentContext != null ? parentContext.db : null, parentContext != null ? parentContext.getConfiguration() : null, null, false);
```
Comment (line-o) [inline exist-core/src/main/java/org/exist/xquery/ModuleContext.java:68]: I find this very hard to read. Can this be refactored for readability?

File: exist-core/src/main/java/org/exist/xquery/XQueryContext.java [hunk 3]
```diff
@@ -3012,12 +3061,10 @@ public Object getAttribute(final String attribute) {
     /**
      * Load the default prefix/namespace mappings table and set up internal functions.
      *
-     * @param config the configuration
+     * @param config the configuration if available
      */
     @SuppressWarnings("unchecked")
-    void loadDefaults(final Configuration config) {
-        this.watchdog = new XQueryWatchDog(this);
```
Comment (line-o) [inline exist-core/src/main/java/org/exist/xquery/XQueryContext.java:3019]: where is the watchdog set now?

File: exist-core/src/test/xquery/xsuite-declare-xqsuite.xqm [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+ : Lesser General Public License for more details.
+ :
+ : You should have received a copy of the GNU Lesser General Public
+ : License along with this library; if not, write to the Free Software
+ : Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ :)
+xquery version "3.1";
+
+(:~
+ : Test for using XQSuite from XSuite via `declare namespace`.
```
Comment (line-o) [inline exist-core/src/test/xquery/xsuite-declare-xqsuite.xqm:25]: XSuite = XQSuite?
If this is just a typo please correct it as it has the potential to confuse people (it did confuse me).

File: exist-core/src/test/xquery/xsuite-import-xqsuite-location.xqm [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ : Lesser General Public License for more details.
+ :
+ : You should have received a copy of the GNU Lesser General Public
+ : License along with this library; if not, write to the Free Software
+ : Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ :)
+xquery version "3.1";
+
+(:~
+ : Test for using XQSuite from XSuite via `import module` with a location hint.
```
Comment (line-o) [inline exist-core/src/test/xquery/xsuite-import-xqsuite-location.xqm:25]: see above

**Review by line-o — COMMENTED** [review]
LGTM. Please have a look at my notes.

File: exist-core/src/test/xquery/xsuite-declare-xqsuite.xqm (hunk 4 above)
Comment (joewiz) [reply exist-core/src/test/xquery/xsuite-declare-xqsuite.xqm:25]: See https://github.com/eXist-db/exist/blob/develop/exist-core/src/main/java/org/exist/test/runner/XSuite.java#L47-L59.

File: exist-core/src/test/resources-filtered/conf.xml (hunk 1 above)
Comment (adamretter) [reply exist-core/src/test/resources-filtered/conf.xml:892]: No.

File: exist-core/src/main/java/org/exist/xquery/ModuleContext.java (hunk 2 above)
Comment (adamretter) [reply exist-core/src/main/java/org/exist/xquery/ModuleContext.java:68]: @line-o That is not possible in Java as a `super` has to be the first call in the constructor!

File: exist-core/src/main/java/org/exist/xquery/XQueryContext.java (hunk 3 above)
Comment (adamretter) [reply exist-core/src/main/java/org/exist/xquery/XQueryContext.java:3019]: You can see it in the main constructor.

File: exist-core/src/test/xquery/xsuite-declare-xqsuite.xqm (hunk 4 above)
Comment (adamretter) [reply exist-core/src/test/xquery/xsuite-declare-xqsuite.xqm:25]: @line-o XSuite is the Java test runner that calls XQSuite. This makes no change to eXide.

File: exist-core/src/test/xquery/xsuite-import-xqsuite-location.xqm (hunk 5 above)
Comment (adamretter) [reply exist-core/src/test/xquery/xsuite-import-xqsuite-location.xqm:25]: A mentioned XSuite != XQSuite

File: exist-core/src/test/xquery/xsuite-import-xqsuite-no-location.xqm [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+ : Lesser General Public License for more details.
+ :
+ : You should have received a copy of the GNU Lesser General Public
+ : License along with this library; if not, write to the Free Software
+ : Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ :)
+xquery version "3.1";
+
+(:~
+ : Test for using XQSuite from XSuite via `import module` with no location hint.
```
[the remark this answers is not in the document]
Comment (adamretter) [reply exist-core/src/test/xquery/xsuite-import-xqsuite-no-location.xqm:25]: A mentioned XSuite != XQSuite

File: exist-core/src/main/java/org/exist/xquery/ModuleContext.java (hunk 2 above)
Comment (line-o) [reply exist-core/src/main/java/org/exist/xquery/ModuleContext.java:68]: I did not know that. How about extracting the logic for the first parameter into a method?

Comment (line-o) [reply exist-core/src/main/java/org/exist/xquery/ModuleContext.java:68]: Maybe reformatting would do the trick?
```java
super(
  parentContext != null ? parentContext.db : null,
  parentContext != null ? parentContext.getConfiguration() : null,
  null,
  false
);
```

Comment (line-o) [reply exist-core/src/main/java/org/exist/xquery/ModuleContext.java:68]: Digging deeper, I wonder if an additional XQueryContext constructor would make sense.
The module context needs to set its XQueryContext here. This is almost like the constructor that copies the context:
https://github.com/eXist-db/exist/blob/cff6e205ff75f7154d037251a584685e4427b4a3/exist-core/src/main/java/org/exist/xquery/XQueryContext.java#L480 

It does not hand over the profiler and never loads defaults, so it is special.
In which cases does the ModuleContext get instantiated without a parentContext?

Comment (line-o) [reply exist-core/src/main/java/org/exist/xquery/ModuleContext.java:68]: What happens when we call `super(parentContext)` instead?

Comment (adamretter) [reply exist-core/src/main/java/org/exist/xquery/ModuleContext.java]: @line-o I have pushed a commit to reformat the call to the super-constructor as you requested.

> I wonder if an additional XQueryContext constructor would make sense.

I don't see how that would help.

> What happens when we call super(parentContext) instead?

It doesn't work.

**Comment by adamretter** [comment]
@line-o (cc @dizzzz) I believe all review has been addressed. Would you mind merging it?

**Pull request merged**